### PR TITLE
Add ConvNext Backbone

### PR DESCRIPTION
**Pull request by gitttt-1234**
## Summary by CodeRabbit

- **New Features**
	- Introduced new classes and enhanced existing ones for improved image processing and inference capabilities, including handling of centroids, confidence maps, and bounding boxes.

### CODE REVIEW

**Comment by talmo** [comment]
Rebased this to `main` now that the other PRs are merged in.

Make sure to:

- [ ] Fix merge conflicts
- [ ] Fix long-running test that is crashing CI

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -9,16 +9,17 @@ on:
...
-  push:
-    branches:
-      - main
-    paths:
-      - "sleap_nn/**"
-      - "tests/**"
-      - ".github/workflows/ci.yml"
-      - "environment_cpu.yml"
-      - "pyproject.toml"
+  # push:
```
Comment (talmo) [inline .github/workflows/ci.yml:14]: Can we uncomment these? I forgot to do that in the upstream PR

File: sleap_nn/architectures/convnext.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import torch
+from torch import nn, Tensor
+from torchvision.ops.misc import Conv2dNormActivation
+from torchvision.utils import _log_api_usage_once
+from sleap_nn.architectures.encoder_decoder import Decoder
+from torchvision.models.convnext import LayerNorm2d, CNBlock, CNBlockConfig
+
+
+class ConvNeXtEncoder(nn.Module):
+    """Src: torchvision.models.
```
Comment (talmo) [inline sleap_nn/architectures/convnext.py:18]: This is not an appropriate class docstring first line. It should be a 1-line summary of the class.

File: sleap_nn/architectures/encoder_decoder.py [hunk 3]
```diff
@@ -462,6 +462,22 @@ def __init__(
...
             current_stride = next_stride
+        while current_stride >= 1:
+            next_stride = current_stride // 2
+            self.decoder_stack.append(
+                SimpleUpsamplingBlock(
+                    x_in_shape=(x_in_shape),
+                    current_stride=current_stride,
+                    upsampling_stride=2,
+                    interp_method="bilinear",
+                    refine_convs=0,  # self.convs_per_block,
```
Comment (talmo) [inline sleap_nn/architectures/encoder_decoder.py:473]: Should this really be hardcoded to 0?

File: sleap_nn/architectures/swint.py [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+)
+
+
+torch.fx.wrap("_patch_merging_pad")
+torch.fx.wrap("_get_relative_position_bias")
+torch.fx.wrap("shifted_window_attention")
+
+
+class SwinTransformerEncoder(nn.Module):
+    """Src: torchvision.models.
```
Comment (talmo) [inline sleap_nn/architectures/swint.py:32]: Not an appropriate class summary (see above)

File: sleap_nn/data/providers.py [hunk 5]
```diff
@@ -48,11 +64,16 @@ def __init__(
 
     @classmethod
     def from_filename(
-        cls, filename: str, user_instances_only: bool = True, max_instances: int = 30
+        cls,
+        filename: str,
+        user_instances_only: bool = True,
+        max_height: int = None,
+        max_width: int = None,
```
Comment (talmo) [inline sleap_nn/data/providers.py:71]:
```suggestion
        max_height: Optional[int] = None,
        max_width: Optional[int] = None,
```

File: sleap_nn/data/providers.py [hunk 6]
```diff
@@ -67,7 +88,8 @@ def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
 
             instances = []
             for inst in lf:
-                instances.append(inst.numpy())
+                if not np.all(np.isnan(inst.numpy())):
```
Comment (talmo) [inline sleap_nn/data/providers.py:91]:
```suggestion
                if not inst.is_empty():
```

File: sleap_nn/data/providers.py [hunk 7]
```diff
@@ -82,6 +104,29 @@ def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
...
+                    )
+                if pad_width < 0:
+                    raise Exception(
+                        f"Max width {self.max_width} should be greater than the current image width: {img_width}"
+                    )
+                image = np.pad(
+                    image,
+                    ((0, 0), (0, 0), (0, pad_height), (0, pad_width)),
+                    mode="constant",
+                ).astype("float32")
```
Comment (talmo) [inline sleap_nn/data/providers.py:125]: I don't love this being here... Is there a reason we can't do padding as a separate, configurable block? [Like we do in core SLEAP](https://github.com/talmolab/sleap/blob/0134ab443663bf1b1f0287d632dce9fdef80c14b/sleap/nn/data/resizing.py#L257-L469)?

File: sleap_nn/data/providers.py [hunk 8]
```diff
@@ -82,6 +104,29 @@ def __iter__(self) -> Iterator[Dict[str, torch.Tensor]]:
...
+
+            # convert to rgb
+            if self.is_rgb and image.shape[-3] != 3:
+                image = np.concatenate([image, image, image], axis=-3)
```
Comment (talmo) [inline sleap_nn/data/providers.py:129]: Ditto the above -- this is an easily factorizable operation. [See core SLEAP version.](https://github.com/talmolab/sleap/blob/0134ab443663bf1b1f0287d632dce9fdef80c14b/sleap/nn/data/normalization.py#L357)

File: sleap_nn/inference/inference.py [hunk 9]
```diff
@@ -165,9 +174,189 @@ def predict(
...
+            threshold=self.peak_threshold,
+            refinement=self.refinement,
+            integral_patch_size=self.integral_patch_size,
+        )
+
+        # Adjust for stride and scale.
+        refined_peaks = refined_peaks * self.output_stride  # (n_centroids, 2)
+        batch, channels, height, width = cms.shape
+
+        batch_indxs = Counter(peak_sample_inds.detach().cpu().numpy().tolist())
```
Comment (talmo) [inline sleap_nn/inference/inference.py:315]:
```suggestion
        batch_inds = Counter(peak_sample_inds.detach().cpu().numpy().tolist())
```

Comment (talmo) [reply sleap_nn/inference/inference.py:315]: And do we really want to move this to the CPU here?

File: sleap_nn/inference/inference.py [hunk 10]
```diff
@@ -245,8 +434,8 @@ def forward(self, batch: Dict[str, torch.Tensor]) -> Dict[str, np.array]:
                 peaks_vals = vals
 
         peaks_output = batch.copy()
-        peaks_output["pred_instance_peaks"] = peaks.unsqueeze(dim=1)
-        peaks_output["pred_peak_values"] = peaks_vals.unsqueeze(dim=1)
+        peaks_output["pred_instance_peaks"] = peaks  # .unsqueeze(dim=1)
+        peaks_output["pred_peak_values"] = peaks_vals  # .unsqueeze(dim=1)
```
Comment (talmo) [inline sleap_nn/inference/inference.py:438]: remove unused comments?

File: sleap_nn/inference/inference.py [hunk 11]
```diff
@@ -165,9 +174,189 @@ def predict(
...
+            current_peak_vals = peak_vals[parsed : parsed + parsed_idx]
+            if len(current_peaks) > self.max_instances:
+                # take top k
+                current_peak_vals, indices = torch.topk(
+                    current_peak_vals, self.max_instances
+                )
+                current_peaks = current_peaks[indices]
+                num_nans = 0
+            else:
+                num_nans = self.max_instances - len(current_peaks)
```
Comment (talmo) [inline sleap_nn/inference/inference.py:333]: There isn't a way we can do this without forcing max instances?

File: sleap_nn/inference/inference.py [hunk 12]
```diff
@@ -229,7 +418,7 @@ def forward(self, batch: Dict[str, torch.Tensor]) -> Dict[str, np.array]:
                     batch_peaks = torch.cat(
                         [
                             batch_peaks,
-                            torch.full((max_inst - num_inst, nodes, 2), torch.nan),
+                            torch.full((abs(max_inst - num_inst), nodes, 2), torch.nan),
```
Comment (talmo) [inline sleap_nn/inference/inference.py:421]: Why abs?? What if `num_inst > max_inst`?

File: sleap_nn/inference/inference.py [hunk 13]
```diff
@@ -431,26 +631,45 @@ class TopDownPredictor(Predictor):
...
             centroid_crop_layer = None
         else:
             self.model_config["centroid"] = self.centroid_config
             self.model_config["data"] = self.centroid_config.inference_config.data
             self.model_config["skeletons"] = self.centroid_config.data_config.skeletons
+            if self.confmap_model:
+                return_crops = True
+            labels = sio.load_slp(
+                self.centroid_config.inference_config.data.labels_path
+            )
```
Comment (talmo) [inline sleap_nn/inference/inference.py:655]: Definitely not. We should 100% not load an SLP just so we can pull out a single scalar parameter!! Store it in the model config!

File: sleap_nn/model_trainer.py [hunk 14]
```diff
@@ -239,14 +289,27 @@ def __init__(self, config: OmegaConf):
...
         self.training_loss = {}
         self.val_loss = {}
         self.learning_rate = {}
 
+        if self.model_config.init_weights == "xavier":
+            self.model.apply(xavier_init_weights)
+
+        if self.model_config.pre_trained_weights:
+            self._init_pretrain_model()
+
```
Comment (talmo) [inline sleap_nn/model_trainer.py:301]: If both of these configs are set, then the pretrained weights takes precedence. If that's the case, then this should be an if-elif block, and the order of precedence should be documented.

File: sleap_nn/model_trainer.py [hunk 15]
```diff
@@ -239,14 +289,27 @@ def __init__(self, config: OmegaConf):
...
+    def _init_pretrain_model(self):
+        if self.model_config.backbone_config.backbone_type in ["swint", "convnext"]:
+            ckpt = eval(self.model_config.pre_trained_weights).DEFAULT.get_state_dict(
+                progress=True, check_hash=True
+            )
+            if self.model_config.backbone_config.backbone_config.in_channels == 1:
+                ckpt["features.0.0.weight"] = torch.unsqueeze(
+                    ckpt["features.0.0.weight"].mean(dim=1), dim=1
+                )
```
Comment (talmo) [inline sleap_nn/model_trainer.py:310]: Interesting decision... I'm not totally opposed to this, but have never seen it before so it makes me a little antsy.

Normally we'd either:
1. Just have a 1x1 conv with a linear projection to get the channels to match up, or
2. Replicate the single channel input image 3x along the channels axis so it's "RGB".

The second is what we do in core SLEAP and would probably be my preference. I'm willing to be convinced otherwise, but I'd prefer to not invent something new for such a fundamental early step of the model processing.

File: tests/architectures/test_model.py [hunk 16]
```diff
@@ -52,7 +116,8 @@ def test_get_head():
 
 
 def test_unet_model():
-    device = "cuda" if torch.cuda.is_available() else "cpu"
+    # device = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (talmo) [inline tests/architectures/test_model.py:119]: ?

File: tests/architectures/test_unet.py [hunk 17]
```diff
@@ -7,7 +7,8 @@
 
 
 def test_unet_reference():
-    device = "cuda" if torch.cuda.is_available() else "cpu"
+    # device = "cuda" if torch.cuda.is_available() else "cpu"
```
Comment (talmo) [inline tests/architectures/test_unet.py:10]: ?

File: tests/data/test_confmaps.py [hunk 18]
```diff
@@ -35,7 +40,7 @@ def test_confmaps(minimal_instance):
     sample = next(iter(datapipe2))
 
     assert sample["confidence_maps"].shape == (1, 2, 50, 50)
-    assert torch.max(sample["confidence_maps"]) == torch.Tensor([0.9867223501205444])
+    assert abs(torch.max(sample["confidence_maps"]) - torch.Tensor([0.9967])) < 1e-4
```
Comment (talmo) [inline tests/data/test_confmaps.py:43]: Prefer `torch.testing.assert_close`

**Review by talmo — CHANGES_REQUESTED** [review]

File: sleap_nn/inference/inference.py (hunk 12 above)
Comment (gitttt-1234) [reply sleap_nn/inference/inference.py:421]: yes I need to add a case when the user doesn't provide max_instances attribute. The above code only works for the case max_instances in not None.

File: sleap_nn/data/providers.py (hunk 8 above)
Comment (gitttt-1234) [reply sleap_nn/data/providers.py:139]: Should we be using torchvision transforms then?

File: sleap_nn/inference/inference.py (hunk 11 above)
Comment (gitttt-1234) [reply sleap_nn/inference/inference.py:333]: I thought it's better to have max_instance set as we need to process all the tensors as a batch

File: sleap_nn/inference/inference.py (hunk 12 above)
Comment (gitttt-1234) [reply sleap_nn/inference/inference.py:434]: Removed `max_instances` dependency and made it Optional. If max_instances is provided, then we set the limit, else all the predicted centroids would be processed

File: sleap_nn/inference/inference.py (hunk 11 above)
Comment (gitttt-1234) [reply sleap_nn/inference/inference.py]: Changed this. max_instances is used only when the user provides or if return_crops = True (when we need to pass it to FindInstancePeaksGroundTruth class)

**Review by talmo — APPROVED** [review]
